### PR TITLE
[SYCL-MLIR]: Adjust use of the Pure interface in the SYCL MLIR dialect

### PR DESCRIPTION
**Pull request by etiotto**
Several operations in the SYCL dialect currently use the `Pure` interface on the operation. The MLIR `Pure` interface is defined as: 

>   // Always speculatable operation that does not touch memory.  These operations
>   // are always legal to hoist or sink.
>   def Pure : TraitList<[AlwaysSpeculatable, NoMemoryEffect]>;

where `NoMemoryEffect` is defined as: 

>  def NoMemoryEffect : MemoryEffects<[]>;

These SYCL operation also define the memory effects on several of their input arguments as `MemRead`. This is a contradiction because the `Pure` interface on an operation asserts that the operation does not access memory. This PR remove the `Pure` interface in those cases.

In addition the PR adds `Pure` to the 'grid' operations in the dialect.

### CODE REVIEW

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td [hunk 1]
```diff
@@ -884,7 +885,8 @@ def SYCLItemGetLinearIDOp
 
 def SYCLNDItemGetGlobalIDOp
     : SYCLMethodOpInterfaceImpl<"nd_item.get_global_id", "NdItemType",
-                                ["get_global_id"], [Pure, SYCLGetID]> {
+                                ["get_global_id"],
```
Comment (Naghasan) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:888]: While I think your patch is correct overall, I'm not too sure about that. By spec this is "pure" as the `nd_item` is immutable and may not even read memory (this function really maps to `SYCLGlobalIDOp`). This applies to `nd_item`, `item`, `group` and `sub_group` but not `id`, `range` and `nd_range`.

Comment (victor-eds) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:888]: > `nd_range`

Isn't `nd_range` also immutable?

Comment (victor-eds) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:888]: I see. If I'm correct, thanks to your LICM pass, we will hoist even if the operation is not pure, right?

**Review by victor-eds — APPROVED** [review]

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td (hunk 1 above)
Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:888]: Correct, as long as there are no operation in the loop that store to the memory that is read by another operation (actually is the store dominates the load we can still hoist both operations).

**Pull request merged**